server.go

```diff
@@ -261,6 +271,160 @@ HANDLE:
+		s.onClosed(frisbeeConn, err)
+		return
+	}
+	if s.ConnContext != nil {
+		connCtx = s.ConnContext(connCtx, frisbeeConn)
+	}
+	wg := new(sync.WaitGroup)
+	closed := atomic.NewBool(false)
+	connCtx, cancel := context.WithCancel(connCtx)
+	handler := func(p *packet.Packet) {
+		handlerFunc := s.handlerTable[p.Metadata.Operation]
+		if handlerFunc != nil {
+			packetCtx := connCtx
+			if s.PacketContext != nil {
+				packetCtx = s.PacketContext(packetCtx, p)
+			}
+			outgoing, action := handlerFunc(packetCtx, p)
+			if outgoing != nil && outgoing.Metadata.ContentLength == uint32(len(*outgoing.Content)) {
+				s.preWrite()
+				err := frisbeeConn.WritePacket(outgoing)
+				if outgoing != p {
+					packet.Put(outgoing)
+				}
+				packet.Put(p)
+				if err != nil {
+					_ = frisbeeConn.Close()
+					if closed.CAS(false, true) {
+						s.onClosed(frisbeeConn, err)
+					}
+					cancel()
+					wg.Done()
+					return
+				}
+			} else {
+				packet.Put(p)
+			}
+			switch action {
+			case NONE:
+			case CLOSE:
+				_ = frisbeeConn.Close()
+				if closed.CAS(false, true) {
+					s.onClosed(frisbeeConn, nil)
+				}
+				cancel()
+			}
+		} else {
+			packet.Put(p)
+		}
+		wg.Done()
+		return
+	}
+HANDLE:
+	wg.Add(1)
+	go handler(p)
+	p, err = frisbeeConn.ReadPacket()
+	if err != nil {
+		_ = frisbeeConn.Close()
+		if closed.CAS(false, true) {
+			s.onClosed(frisbeeConn, err)
+		}
+		cancel()
+		wg.Wait()
+		return
+	}
+	goto HANDLE
+}
+
+func (s *Server) handleLimitedPacket(frisbeeConn *Async, connCtx context.Context) {
+	p, err := frisbeeConn.ReadPacket()
+	if err != nil {
+		_ = frisbeeConn.Close()
+		s.onClosed(frisbeeConn, err)
+	}
+	if s.ConnContext != nil {
+		connCtx = s.ConnContext(connCtx, frisbeeConn)
+	}
+	wg := new(sync.WaitGroup)
+	closed := atomic.NewBool(false)
+	connCtx, cancel := context.WithCancel(connCtx)
+	concurrency := make(chan struct{}, s.concurrency)
+	handler := func(p *packet.Packet) {
+		handlerFunc := s.handlerTable[p.Metadata.Operation]
+		if handlerFunc != nil {
+			packetCtx := connCtx
+			if s.PacketContext != nil {
+				packetCtx = s.PacketContext(packetCtx, p)
+			}
+			outgoing, action := handlerFunc(packetCtx, p)
+			if outgoing != nil && outgoing.Metadata.ContentLength == uint32(len(*outgoing.Content)) {
+				s.preWrite()
+				err := frisbeeConn.WritePacket(outgoing)
+				if outgoing != p {
+					packet.Put(outgoing)
+				}
+				packet.Put(p)
+				if err != nil {
+					_ = frisbeeConn.Close()
+					if closed.CAS(false, true) {
+						s.onClosed(frisbeeConn, err)
+					}
+					cancel()
+					<-concurrency
+					wg.Done()
+					return
+				}
+			} else {
+				packet.Put(p)
+			}
+			switch action {
+			case NONE:
+			case CLOSE:
+				_ = frisbeeConn.Close()
+				if closed.CAS(false, true) {
+					s.onClosed(frisbeeConn, nil)
+				}
+				cancel()
+			}
+		} else {
+			packet.Put(p)
+		}
+		<-concurrency
+		wg.Done()
+		return
+	}
+HANDLE:
+	select {
+	case concurrency <- struct{}{}:
+		wg.Add(1)
+		go handler(p)
```

Comment:
The handler here is pretty much identical to the one for unlimited. I'd share the code in s.handlePacket() then do something like:

go func(packet) {
    s.handlePacket(packet, wg)
    <-concurrency
}(p)

Reply:
Makes sense!